### ISSUE TITLE
FILENAME bound variable?

### ISSUE THREAD

**Issue by chippydip** [KSP-KOS/KOS#1946]
Not sure if this is just an omission in the documentation or something weird going on, but it certainly surprised me and I couldn't find it documented anywhere.

Calling a script from another script:

    RUNPATH("0:/trajectory_opt.ks", stages, fileName).

and then grabbing the parameters in that script:

    parameter stages.
    parameter fileName.

doesn't actually pass in the second parameter. Instead, it contains the filename of the script being called ("trajectory_opt" in this case). If I print out `fileName` before the parameter declarations it already has that value and the parameter declaration doesn't modify it's value (but it does pop the argument off the stack, because if I add a 3rd parameter after that I get a parameter list length error).

This is the work-around I ended up with:

    parameter stages.
    parameter tmp.
    set fileName to tmp.

**Comment by Dunbaratu** [comment]
What value was ``fileName`` supposed to have?  Can you print it out just before the RUNPATH line, like so?
```
print "Before RUNPATH, value of fileName is: " + fileName.
RUNPATH("0:/trajectory_opt.ks", stages, fileName).
```
Then inside the program:
```
parameter stages.
parameter fileName.
print "Inside trajectory_opt, value of fileName is: " + fileName.
```
Without seeing that information, it's hard to diagnose what's happening.

**Comment by chippydip** [comment]
Yeah, I did exactly that when debugging the issue. Here's a minimal repro case:

test1.ks:
    RUNPATH("0:/test2.ks", "foo").

test2.ks:
    print "test2 pre-params: " + fileName.
    parameter fileName.
    print "test2 post-params: " + fileName.

output:
    test2 pre-params: 0:/test2.ks
    test2 post-params: 0:/test2.ks

It's also possible this is something that's been fixed already. I'm playing RSS/RP-0 so running KSP 1.1.3 and kOS 1.0.1 (sorry, should have mentioned that in the initial report).

**Comment by Dunbaratu** [comment]
That isn't the test I meant.

This is the test I meant:

test1.ks:
```
print "BEFORE executing runpath, what is the value of filename?".
print "fileName is " + fileName.
RUNPATH("0:/test2.ks", "foo").
```
What I'm asking is whether or not there's *already* a variable called fileName that existed prior to using it as the name of a parameter.  I'm trying to figure out if there is, and if so whether the scoping rules are messing up and seeing that one instead of the parameter.

**Comment by chippydip** [comment]
As expected, trying to run that version of `test1.ks` throws an undefined variable error when trying to print `fileName`.

I had a hunch, though, and played around with how the second script is being called:

* `RUNPATH("0:/test2.ks", "foo").` -> "0:/test2.ks"
* `RUNPATH("test2.ks", "foo").` -> "test2.ks"
* `RUNPATH("test2", "foo").` -> "test2"
* `run test2("foo").` -> "test2" (so, it's not even specific to `RUNPATH`)

So, it seems like there's a `FILENAME` variable that gets set to whatever path the file was executed as. This would be fine, except that trying to use `parameter FILENAME` seems to pop the argument off the stack into some scope that's hidden by the implicitly defined `FILENAME` that was set when the file was run.

**Comment by Dunbaratu** [comment]
Thanks for the in-depth information.  That should make it easier to track down where this is happening.  I suspect the program running function (that checks to see if the program was already compiled) is using it's own internal global variable that needs to be renamed from "filename" to something ridiculously absurd that nobody could use as a normal identifier, something like ``file''name''``.

**Comment by jinks** [comment]
Python uses `__variableName` for its internal, hardcoded stuff. Would that be an option here? This way users *can* find it if they really want to for... reasons.

**Comment by Tricky1702** [comment]
Try scriptpath().

**Comment by jinks** [comment]
@Tricky1702 What does that have to do with renaming an internal kOS variable?

**Comment by Tricky1702** [comment]
The title might want re-wording.
